[distributed] quicker exit in the case of failed tests in distributed

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34413 [rpc] improve general error handling in process group agent and fix flaky tests
* **#34150 [distributed] quicker exit in the case of failed tests in distributed**

In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

rohan-varma committed:
[distributed] quicker exit in the case of failed tests in distributed

In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful "Process 0 timed out..." error
message when the error was actually something else.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -270,6 +295,21 @@ def _check_return_codes(self, elapsed_time):
         first_process = self.processes[0]
+        # first, we check if there are errors in actual processes (via TEST_ERROR_EXIT CODE), and raise an exception for those.
+        # the reason we do this is to attempt to raise a more helpful error message than "Process x terminated/timed out"
+        # TODO: we should pipe the exception of the failed subprocess here. Currently, the actual exception is displayed as a log output.
```
Let's break these lines into shorter pieces.

mrshenli commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -247,8 +247,33 @@ def _run(cls, rank, test_name, file_name):
-        for p in self.processes:
-            p.join(timeout)
+        subprocess_error = False
+        while True:
```
This is a busy loop? do we need to insert some short sleeps in it?

rohan-varma replied:
Yes, I guess we should insert some sleeps. Although if our sleep interval is too long we could lose some benefit of being able to return quickly in the case of errors. Ideally this process would use a pipe or condvar of some sort that just sleeps until timeout or until one process exits.

mrshenli replied:
If you would like to land this sooner, a 10ms sleep should also be fine. We already have this in our test. Totally up to you :)

https://github.com/pytorch/pytorch/blob/24dd800e6a1471e595099dc9fdd48c19c71a4e20/torch/testing/_internal/distributed/rpc/dist_autograd_test.py#L51-L53

https://github.com/pytorch/pytorch/blob/24dd800e6a1471e595099dc9fdd48c19c71a4e20/torch/testing/_internal/distributed/rpc/dist_autograd_test.py#L1366-L1368

https://github.com/pytorch/pytorch/blob/24dd800e6a1471e595099dc9fdd48c19c71a4e20/torch/testing/_internal/distributed/rpc/rpc_test.py#L1413-L1421

mrshenli commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -247,8 +247,33 @@ def _run(cls, rank, test_name, file_name):
+                # This is the exited code processes exit with if they encountered an exception.
+                if p.exitcode == MultiProcessTestCase.TEST_ERROR_EXIT_CODE:
+                    print("Some process exited badly, terminating rest.")
+                    active_children = torch.multiprocessing.active_children()
```
Would I be correct if I assume this only returns living sub-processes launched by the current process?

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -270,6 +295,21 @@ def _check_return_codes(self, elapsed_time):
+            if p.exitcode == MultiProcessTestCase.TEST_ERROR_EXIT_CODE
+        ]
+        if errored_processes:
+            error = "Processes {} errored exited with error code {}".format(
```
errored exited -> exited?

mrshenli commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -247,8 +247,33 @@ def _run(cls, rank, test_name, file_name):
+                    )
+                )
+                for p in self.processes:
+                    p.terminate()
```
will this be a no-op for errored processes? Because there could be race that some process crashed after we check their error code above.

rohan-varma replied:
Yes, it looks like it is safe to call `terminate()` multiple times/on an already-exited process. Although, maybe we could use `multiprocessing.active_children()`.

mrshenli reviewed (approved):
LGTM! Left some minor comments. Stamp to unblock. Thanks!

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

xush6528 commented:
Just for FYI, 
I remember `torch.mp.spawn(..)` use `mp.sentinal`
https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.sentinel

https://github.com/pytorch/pytorch/blob/7314f1c2818978fb28b420710ffe1f3b1c0b95a1/torch/multiprocessing/spawn.py#L75-L79

rohan-varma commented:
@xush6528 Thanks for flagging, i will look into that and see why it doesn't seem to be working as expected

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] quicker exit in the case of failed tests in distributed"


In the distributed setting we commonly have tests in which there are errors where one process
exits but the other do not (since they are for example waiting for work from
the process that exited). Currently, when this situation happens we do not
handle this well, and wait for process 0 to timeout. This results in wasted
time waiting for test errors and a less helpful `"Process 0 timed out..."` error
message when the error was actually something else. See https://github.com/pytorch/pytorch/issues/34149 for more details.

This diff fixes the issue by checking for exited subprocesses and terminating
the test when we see a subprocess that has exited uncleanly. We still enforce
timeouts and return when all processes have exited cleantly in the happy path.

Tested by ensuring that if we have an uncaught error in one rank, we quickly terminate the test, previously, we would wait for timeout.

Differential Revision: [D20231032](https://our.internmc.facebook.com/intern/diff/D20231032/)

[ghstack-poisoned]

rohan-varma commented:
ROCm failure looks due to a network issue:

```
05:52:36 ++ sccache --stop-server
05:52:36 Stopping sccache server...
05:52:36 error: couldn't connect to server
05:52:36 caused by: Connection refused (os error 111)
```
